action-allowedlist/action_allowedlist/actions_parser.py

```diff
@@ -0,0 +1,116 @@
+# SPDX-License-Identifier: Apache-2.0
+# Licensed to the Ed-Fi Alliance under one or more agreements.
+# The Ed-Fi Alliance licenses this file to you under the Apache License, Version 2.0.
+# See the LICENSE and NOTICES files in the project root for more information.
+
+import json
+import yaml
+from pathlib import Path
+
+
+def load_json_file(filepath):
+    with open(filepath, 'r') as file:
+        content = file.read()
+        return json.loads(content)
+
+def load_yaml_file(filepath):
+    with open(filepath, 'r') as file:
+        return yaml.safe_load(file)
+
+def get_actions_from_file(workflow, workflow_file_name):
+    #print(workflow_file_name)
+    parsed_yaml = yaml.safe_load(workflow)
+    actions = []
+
+    for job_name, job in parsed_yaml.get("jobs", {}).items():
+        print(f"  Job found: [{job_name}] in {workflow_file_name}")
+        steps = job.get("steps", [])
+        for step in steps:
+            uses = step.get("uses")
+            if (uses != None):
+                parts = uses.split("@")
+                if len(parts) == 2:
+                    action_link, action_version = parts
+                    actions.append({
+                        "actionLink": action_link,
+                        "actionVersion": action_version,
+                        "workflowFileName": workflow_file_name
+                    })
+                    print(f"   Found action used: [{uses}]")
+
+    return actions
+
+def get_all_used_actions(workflow_dir: Path):
+    print("Loading Actions YAML files")
+    if (workflow_dir / ".github/workflows").exists():
+        workflow_files = list((workflow_dir / ".github/workflows").glob("*.yml"))
+    else:
+        workflow_files = list((workflow_dir / "testing-repo/.github/workflows").glob("*.yml"))
+    if not workflow_files:
+        print("Could not find workflow files in the specified directory")
+        return []
+
+    print(f"Found [{len(workflow_files)}] files in the workflows directory")
+    #print(workflow_files)
+    actions_in_repo = []
+    for workflow_file in workflow_files:
+        #print(workflow_file)
+        try:
+            workflow_content = workflow_file.read_text()
+            actions = get_actions_from_file(workflow_content, workflow_file.name)
+            actions_in_repo.extend(actions)
+        except Exception as e:
+            print(f"Error occurred while reading {workflow_file}: {e}")
+
+    return actions_in_repo
+
+def invoke_validate_actions(approved_path, actions_configuration):
+    print("Checking if used actions are approved")
+
+    approved = load_json_file(approved_path)
+    num_approved = 0
+    num_denied = 0
+    num_deprecated = 0
+    unapproved_outputs = []
+    approved_outputs = []
+    found = False
+
+
+    for action in actions_configuration:
+        # print(f"Processing {action['actionLink']} version {action['actionVersion']}") Use for debugging
```

Comment:
You can uncomment this and modify to print to debug. See [Setting a debug message](https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions).

```
print(f"::debug::Processing {action['actionLink']} version {action['actionVersion']}")
```